src/analysis/locally_convex/weak_dual.lean

```diff
@@ -0,0 +1,149 @@
+/-
+Copyright (c) 2022 Moritz Doll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Doll
+-/
+import topology.algebra.module.weak_dual
+import analysis.normed.normed_field
+import analysis.locally_convex.with_seminorms
+
+/-!
+# Weak Dual in Topological Vector Spaces
+
+## Main definitions
+
+* `weak_bilin_basis_zero`: a basis for the neighborhood filter at 0.
+* `linear_map.to_seminorm_family`: turn a bilinear form `B : E →ₗ[𝕜] F →ₗ[𝕜] 𝕜`
+into a map `F → seminorm 𝕜 E`.
+
+## Main statements
+
+* `with_seminorms B.to_seminorm_family`: the topology of a weak space is induced by the family of
+seminorm `B.to_seminorm_family`.
+* `weak_bilin.to_locally_convex_space'`: a spaced endowed with a weak topology is locally convex.
+
+## References
+
+* [Bourbaki, *Topological Vector Spaces*][bourbaki1987]
+
+## Tags
+
+weak dual, seminorm
+-/
+
+variables {𝕜 E F ι : Type*}
+
+open_locale topological_space
+
+section bilin_form
+
+namespace linear_map
+
+variables [normed_field 𝕜] [add_comm_group E] [module 𝕜 E] [add_comm_group F] [module 𝕜 F]
+
+/-- Construct a seminorm from a linear form `f : E →ₗ[𝕜] 𝕜` over a normed field `𝕜` by
+`λ x, ∥f x∥` -/
+def to_seminorm (f : E →ₗ[𝕜] 𝕜) : seminorm 𝕜 E :=
+{ to_fun := λ x, ∥f x∥,
+  smul' := λ a x, by simp only [map_smulₛₗ, ring_hom.id_apply, smul_eq_mul, norm_mul],
+  triangle' := λ x x', by { simp only [map_add, add_apply], exact norm_add_le _ _ } }
+
+lemma coe_to_seminorm {f : E →ₗ[𝕜] 𝕜} :
+  ⇑f.to_seminorm = λ x, ∥f x∥ := rfl
+
+@[simp] lemma to_seminorm_apply {f : E →ₗ[𝕜] 𝕜} {x : E} :
+  f.to_seminorm x = ∥f x∥ := rfl
+
+lemma to_seminorm_ball_zero {f : E →ₗ[𝕜] 𝕜} {r : ℝ} :
+  seminorm.ball f.to_seminorm 0 r = { x : E | ∥f x∥ < r} :=
+by simp only [seminorm.ball_zero_eq, to_seminorm_apply]
+
+lemma to_seminorm_comp (f : F →ₗ[𝕜] 𝕜) (g : E →ₗ[𝕜] F) :
+  f.to_seminorm.comp g = (f.comp g).to_seminorm :=
+by { ext, simp only [seminorm.comp_apply, to_seminorm_apply, coe_comp] }
+
+/-- Construct a family of seminorms from a bilinear form. -/
+def to_seminorm_family (B : E →ₗ[𝕜] F →ₗ[𝕜] 𝕜) : seminorm_family 𝕜 E F :=
+λ y, (B.flip y).to_seminorm
+
+@[simp] lemma to_seminorm_family_apply {B : E →ₗ[𝕜] F →ₗ[𝕜] 𝕜} {x y} :
+  (B.to_seminorm_family y) x = ∥B x y∥ := rfl
+
+/-- A basis of the weak topology `weak_bilin B`. -/
+def weak_bilin_basis_zero (B : E →ₗ[𝕜] F →ₗ[𝕜] 𝕜) : set (set E) :=
+⋃ (s : finset F) (hs : s.nonempty) r (hr : 0 < r), { s.inf' hs (λ y, { x : E | ∥B x y∥ < r}) }
```

Comment:
I don't really understand the point of this definition: it is not used anywhere (in particular, it is not shown to be a basis for the neighborhood filter at zero as advertised), and it seems that further down in `has_basis_weak_bilin` there is a different basis that is used for this anyway.

Reply:
Oh no, I forgot to delete old code..